### PR TITLE
New default KDF

### PR DESCRIPTION
**Pull request by L11R**
HKDF-HMAC-SHA256 instead of just SHA256

### CODE REVIEW

**Comment by kigawas** [comment]
Can you add some test cases?

**Comment by L11R** [comment]
@kigawas it tests against your Python backend, I cannot fix it.

**Comment by kigawas** [comment]
https://github.com/ecies/py/pull/87

**Comment by L11R** [comment]
@kigawas did you test it with JS one?

**Comment by kigawas** [comment]
I'll test it later today

**Comment by kigawas** [comment]
```python
    def test_hdkf(self):
        k1 = PrivateKey(secret=bytes([2]))
        self.assertEqual(k1.to_int(), 2)

        k2 = PrivateKey(secret=bytes([3]))
        self.assertEqual(k2.to_int(), 3)

        self.assertEqual(
            encapsulate(k1, k2.public_key), decapsulate(k1.public_key, k2)
        )
        self.assertEqual(
            encapsulate(k1, k2.public_key).hex(),
            '6f982d63e8590c9d9b5b4c1959ff80315d772edd8f60287c9361d548d5200f82'
        )
```

**Comment by L11R** [comment]
@kigawas if you were surprised with encapsulate/decapsulate syntax, I took it from google's tink, because we need same public key. KEM is just key encapsulation mechanism. ECIES is basically KEM+DEM (data encapsulation mechanism).

**Comment by L11R** [comment]
@L11R let me know when you will have deployed new Python backend at Heroku. I want to test Go version against it.

**Comment by kigawas** [comment]
@L11R 
Already deployed

**Comment by L11R** [comment]
@kigawas if so, JS tests fail :\
AES-GCM cannot decrypt data.
`(node:109) UnhandledPromiseRejectionWarning: Error: Unsupported state or unable to authenticate data`

**Comment by kigawas** [comment]
I'll fix js version, wait a sec

**Comment by kigawas** [comment]
Looks ok for now.
Basically the reason is `ecdhUnsafe` was returning compressed public key format

File: src/keys/PrivateKey.ts [hunk 1]
```diff
@@ -24,8 +25,17 @@ export default class PrivateKey {
...
+        return hkdf(Buffer.concat([
+            this.publicKey.uncompressed,
+            this.multiply(pub),
+        ]), 32, {
+                hash: "SHA-256",
+            });
+    }
+
+    public multiply(pub: PublicKey): Buffer {
+        return secp256k1.ecdhUnsafe(pub.compressed, this.secret, false);
```
Comment (kigawas) [inline src/keys/PrivateKey.ts]: `false` means uncompressed format

**Comment by L11R** [comment]
@kigawas yeah, already got it. Will fix Go version soon.

File: src/keys/PublicKey.ts [hunk 2]
```diff
@@ -31,6 +33,15 @@ export default class PublicKey {
         }
     }
 
+    public decapsulateKEM(priv: PrivateKey): Buffer {
```
Comment (kigawas) [inline src/keys/PublicKey.ts]: can we remove `KEM`?

**Comment by kigawas** [comment]
```python
        derived = HKDF(b'secret', 32, b'', SHA256).hex()
        self.assertEqual(
            derived,
            "2f34e5ff91ec85d53ca9b543683174d0cf550b60d5f52b24c97b386cfcf6cbbf"
        )
```
You can use this to check hkdf

**Comment by L11R** [comment]
@kigawas yeah, already done.

https://github.com/ecies/go/pull/3

File: src/keys/PublicKey.ts (hunk 2 above)
Comment (L11R) [reply src/keys/PublicKey.ts]: @kigawas sure, but I guess we should at least comment it.

**Pull request merged**